actsync/actionsync.go

```diff
@@ -0,0 +1,199 @@
+	if !as.IsReady() {
+		return
+	}
+	// check if the action is already requested
+	_, ok := as.actions.LoadOrStore(hash, &actionMsg{})
+	if ok {
+		log.L().Debug("Action already requested", log.Hex("hash", hash[:]))
+		return
+	}
+	log.L().Debug("Requesting action", log.Hex("hash", hash[:]))
+	as.trigger(hash)
+	return
+}
+
+// ReceiveAction receives an action
+func (as *ActionSync) ReceiveAction(_ context.Context, hash hash.Hash256) {
+	if !as.IsReady() {
+		return
+	}
+	log.L().Debug("received action", log.Hex("hash", hash[:]))
+	as.actions.Delete(hash)
+}
+
+func (as *ActionSync) sync() {
+	defer as.wg.Done()
+	for hash := range as.syncChan {
+		log.L().Debug("syncing action", log.Hex("hash", hash[:]))
+		ctx, cancel := context.WithTimeout(context.Background(), unicaseTimeout)
+		defer cancel()
+		msg, ok := as.actions.Load(hash)
+		if !ok {
+			log.L().Debug("action not requested or already received", log.Hex("hash", hash[:]))
+			continue
+		}
+		if time.Since(msg.(*actionMsg).lastTime) < as.cfg.Interval {
+			log.L().Debug("action is recently requested", log.Hex("hash", hash[:]))
+			continue
+		}
+		msg.(*actionMsg).lastTime = time.Now()
+		if err := as.requestFromNeighbors(ctx, hash); err != nil {
+			log.L().Warn("Failed to request action from neighbors", zap.Error(err))
+		}
+	}
+	log.L().Info("quitting action sync")
+}
+
+func (as *ActionSync) triggerSync() {
+	defer as.wg.Done()
+	ticker := time.NewTicker(as.cfg.Interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			as.actions.Range(func(key, value interface{}) bool {
+				as.trigger(key.(hash.Hash256))
+				return true
+			})
+		case <-as.quit:
+			log.L().Info("quitting action trigger sync")
+			return
+		}
+	}
+}
+
+func (as *ActionSync) trigger(hash hash.Hash256) {
+	if !as.IsReady() {
+		return
+	}
+
+	select {
+	case as.syncChan <- hash:
+	default:
+		log.L().Warn("action sync channel is full, fail to sync action", log.Hex("hash", hash[:]))
+	}
+}
+
+func (as *ActionSync) selectPeers() ([]peer.AddrInfo, error) {
+	neighbors, err := as.helper.P2PNeighbor()
+	if err != nil {
+		return nil, err
+	}
+	repeat := batchPeerSize
+	if repeat > len(neighbors) {
+		repeat = len(neighbors)
+	}
+	if repeat == 0 {
+		return nil, errors.New("no peers")
+	}
+	peers := make([]peer.AddrInfo, repeat)
+	for i := 0; i < repeat; i++ {
+		peer := neighbors[fastrand.Uint32n(uint32(len(neighbors)))]
+		peers[i] = peer
+	}
+	return peers, nil
+}
+
+func (as *ActionSync) requestFromNeighbors(ctx context.Context, hash hash.Hash256) error {
+	l := log.L().With(log.Hex("hash", hash[:]))
+	neighbors, err := as.selectPeers()
+	if err != nil {
+		l.Debug("Failed to get neighbors", zap.Error(err))
+		return err
+	}
+	success := false
+	for i := range neighbors {
+		if err := as.helper.UnicastOutbound(ctx, neighbors[i], &iotexrpc.ActionSync{Hashes: [][]byte{hash[:]}}); err != nil {
+			l.Debug("Failed to request action", zap.Error(err), zap.String("peer", neighbors[i].String()))
```

Comment:
`ActionSync` is designed to contain multiple hash/actions, so can use 1 message to request more than 1 actions per outbound unicast. Should consider this to enhance the network efficiency?

Reply:
good idea, can impl it in next pr

Reply:
add a note/TODO